streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -53,7 +54,7 @@
 import static org.apache.kafka.common.utils.Utils.getPort;
 import static org.apache.kafka.streams.processor.internals.InternalTopicManager.WINDOW_CHANGE_LOG_ADDITIONAL_RETENTION_DEFAULT;
 
-public class StreamPartitionAssignor implements PartitionAssignor, Configurable {
+public class StreamPartitionAssignor implements PartitionAssignor, Configurable, TaskIdToPartitionsProvider {
```

Comment:
One general comment: current there is a mutual dependency between `StreamPartitionAssignor` and `StreamThread` that each of the object needs to maintain a reference of each other. Making them extending an interface to leak as little functionalities as possible is better but with `TaskManager` we still have a circle dependency or `task manager -> partition assignor -> thread data provider (stream thread) -> task manager`.

I'm wondering if we can break this circular dependency. One idea is that we can break partition assignor's dependency on stream thread:

1. builder / name / processorId / config / partitionGrouper are static and won't change after being constructed. We can pass them once in `configure`  from StreamThread when constructing the consumer, but not passing in the `ThreadDataProvider aka StreamThread` itself.

2. `prevActiveTasks` and `cachedTasks` are from `task manager`, we can let `task manager` to hold a reference of the `assignor` since it is dependent on that anyways (so we can replace `setPartitionAssignor` as well), so that it can update these two fields whenever necessary.

So that the dependency now becomes `stream thread -> task manager -> partition assignor`.